File: NickvisionSpotlight/Models/SpotlightManager.cpp
#include "SpotlightManager.h"
#include <stdexcept>
#include <windows.h>

namespace NickvisionSpotlight::Models
{
	SpotlightManager::SpotlightManager() : m_spotlightDir(std::string(getenv("localappdata")) + "\\Packages\\Microsoft.Windows.ContentDeliveryManager_cw5n1h2txyewy\\LocalState\\Assets\\"), m_dataDir(std::string(getenv("appdata")) + "\\Nickvision\\NickvisionSpotlight\\Images\\")
	{
		if (!std::filesystem::exists(m_spotlightDir))
		{
			throw std::runtime_error("Unable to access Windows Spotlight images directory.");
		}
		if (!std::filesystem::exists(m_dataDir))
		{
			std::filesystem::create_directories(m_dataDir);
		}
	}

	const std::vector<std::filesystem::path>& SpotlightManager::GetSpotlightImages() const
	{
		return m_spotlightImages;
	}

	void SpotlightManager::SyncSpotlightImages()
	{
		m_spotlightImages.clear();
		for (const std::filesystem::path& path : std::filesystem::directory_iterator(m_spotlightDir))
		{
			if (std::filesystem::file_size(path) / 1000 >= 200)
			{
				std::filesystem::path newPath = std::filesystem::path(m_dataDir.string() + path.stem().string() + ".jpg");
				std::filesystem::copy(path, newPath, std::filesystem::copy_options::overwrite_existing);
				m_spotlightImages.push_back(newPath);
			}
		}
	}

	void SpotlightManager::ExportImage(size_t selectedPath, const std::filesystem::path& savePath) const
	{
		try
		{
			std::filesystem::copy(m_spotlightImages.at(selectedPath), savePath, std::filesystem::copy_options::overwrite_existing);
		}
		catch (...)
		{
			throw std::invalid_argument("Invalid path index");
		}
	}

	void SpotlightManager::ExportAllImages(const std::filesystem::path& saveDir) const
	{
		for (const std::filesystem::path& path : m_spotlightImages)
		{
			std::filesystem::path copyPath = std::filesystem::path(saveDir.string() + path.filename().string());
			std::filesystem::copy(path, copyPath, std::filesystem::copy_options::overwrite_existing);
		}
	}

	void SpotlightManager::SetImageAsBackground(size_t selectedPath) const
	{
		std::filesystem::path backgroundPath = std::filesystem::path(m_dataDir.string() + "background.jpg");
		try
		{
			std::filesystem::copy(m_spotlightImages.at(selectedPath), backgroundPath, std::filesystem::copy_options::overwrite_existing);
		}
		catch (...)
		{
			throw std::invalid_argument("Invalid path index");
		}
		std::string backgroundPathAsString = backgroundPath.string();
		SystemParametersInfo(SPI_SETDESKWALLPAPER, 0, (void*)(std::wstring(backgroundPathAsString.begin(), backgroundPathAsString.end()).c_str()), SPIF_UPDATEINIFILE);
	}
}

File: NickvisionSpotlight.Shared/Models/SpotlightManager.cs
using NickvisionSpotlight.Shared.Models;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Threading.Tasks;
using Vanara.PInvoke;

namespace NickvisionSpotlight.Models;

/// <summary>
/// A class for managing Windows Spotlight images
/// </summary>
public class SpotlightManager
{
    private string _spotlightDir;
    private string _dataDir;

    /// <summary>
    /// A list of paths of cached spotlight images
    /// </summary>
    public List<string> SpotlightImages { get; init; }

    /// <summary>
    /// Occurs when the list of spotlight images is changed
    /// </summary>
    public event EventHandler<EventArgs>? ImagesChanged;

    /// <summary>
    /// Constructs a SpotlightManager
    /// </summary>
    /// <exception cref="ApplicationException">Thrown if the Windows Spotlight folder does not exist</exception>
    public SpotlightManager()
    {
        _spotlightDir = $"{Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData)}\\Packages\\Microsoft.Windows.ContentDeliveryManager_cw5n1h2txyewy\\LocalState\\Assets\\";
        _dataDir = $"{Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)}{Path.DirectorySeparatorChar}Nick vision{Path.DirectorySeparatorChar}{AppInfo.Current.Name}{Path.DirectorySeparatorChar}Images{Path.DirectorySeparatorChar}";
        SpotlightImages = new List<string>();
        if (!Directory.Exists(_spotlightDir))
        {
            throw new ApplicationException("Unable to find Windows Spotlight images directory.");
        }
        if (!Directory.Exists(_dataDir))
        {
            Directory.CreateDirectory(_dataDir);
        }
    }

    /// <summary>
    /// Scans the Windows Spotlight folder for images and adds them to the Nickvision Spotlight cached images folder, while populating the SpotlightImages list
    /// </summary>
    public async Task SyncSpotlightImagesAsync()
    {
        SpotlightImages.Clear();
        await Task.Run(() =>
        {
            foreach (var path in Directory.EnumerateFiles(_spotlightDir))
            {
                if (new FileInfo(path).Length / 1000 >= 200)
                {
                    var newPath = $"{_dataDir}{Path.GetFileNameWithoutExtension(path)}.jpg";
                    File.Copy(path, newPath, true);
                }
            }
            foreach (var path in Directory.EnumerateFiles(_dataDir))
            {
                using var image = Image.FromFile(path);
                if (image.Width > image.Height)
                {
                    SpotlightImages.Add(path);
                }
                else
                {
                    image.Dispose();
                    File.Delete(path);
                }
            }
        });
        ImagesChanged?.Invoke(this, EventArgs.Empty);
    }

    /// <summary>
    /// Exports a spotlight image
    /// </summary>
    /// <param name="index">The index of the image to export</param>
    /// <param name="savePath">The path of where to export the image to</param>
    /// <exception cref="IndexOutOfRangeException">Thrown if index is invalid</exception>
    public void ExportImage(int index, string savePath)
    {
        if (index >= 0 && index < SpotlightImages.Count)
        {
            File.Copy(SpotlightImages[index], savePath, true);
        }
        else
        {
            throw new IndexOutOfRangeException("The index provided is out of bounds of the SpotlightImages list.");
        }
    }

    /// <summary>
    /// Exports all spotlight images to a directory
    /// </summary>
    /// <param name="saveDir">The directory to export all images</param>
    public async Task ExportAllImagesAsync(string saveDir)
    {
        await Task.Run(() =>
        {
            foreach (var path in SpotlightImages)
            {
                var copyPath = $"{saveDir}{Path.DirectorySeparatorChar}{Path.GetFileName(path)}";
                File.Copy(path, copyPath, true);
            }
        });
    }

    /// <summary>
    /// Sets a spotlight image as the desktop background
    /// </summary>
    /// <param name="index">The index of the image to use as the desktop background</param>
    /// <returns>True, unless there is an error with SystemParametersInfo</returns>
    /// <exception cref="IndexOutOfRangeException">Thrown if index is invalid</exception>
    public bool SetImageAsDesktopBackground(int index)
    {
        if (index >= 0 && index < SpotlightImages.Count)
        {
            var backgroundPath = $"{Environment.GetFolderPath(Environment.SpecialFolder.MyPictures)}{Path.DirectorySeparatorChar}background.jpg";
            File.Copy(SpotlightImages[index], backgroundPath, true);
            return User32.SystemParametersInfo(User32.SPI.SPI_SETDESKWALLPAPER, 0, backgroundPath, User32.SPIF.SPIF_UPDATEINIFILE | User32.SPIF.SPIF_SENDWININICHANGE);
        }
        else
        {
            throw new IndexOutOfRangeException("The index provided is out of bounds of the SpotlightImages list.");
        }
    }
}